Automatic GROUP/ORDER BY to disk based on the memory usage

azat opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Automatic `GROUP BY`/`ORDER BY` to disk based on the server/user memory usage. Controlled with `max_bytes_ratio_before_external_group_by`/`max_bytes_ratio_before_external_sort` query settings.

Fixes: #69286

azat committed:
Add annotation for calling constructor of Aggregator::Params

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Automatic GROUP BY to disk (max_bytes_ratio_before_external_group_by)

Simpler then max_bytes_before_external_group_by, that allows you not to
think about memory consumption of the query, and can be set globally to
i.e. 0.5 so that the external aggregation will be enabled once the
memory usage reaches the 50%.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Introduce getCurrentQueryMemoryUsage() helper for Aggregator

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #71406 from azat/ClickHouse:automatic-external-aggregation -->
*This is an automated comment for commit a2738bb81311f747b97cf20cb4af223abfe681d2 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71406/a2738bb81311f747b97cf20cb4af223abfe681d2/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/71402/5c69cf3205c77bbb2dcf8a5d52539679507c86fe/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

novikd commented on src/Core/Settings.cpp:
```diff
@@ -2302,6 +2302,9 @@ What to do when the limit is exceeded.
     DECLARE(UInt64, max_bytes_before_external_group_by, 0, R"(
 If memory usage during GROUP BY operation is exceeding this threshold in bytes, activate the 'external aggregation' mode (spill data to disk). Recommended value is half of the available system memory.
+)", 0) \
+    DECLARE(Double, max_bytes_ratio_before_external_group_by, 0., R"(
```
Requires adding it to the documentation.

azat replied:
Done [b72e705](https://github.com/ClickHouse/ClickHouse/pull/71406/commits/b72e70576781667b0cc9d326f4cbd33584bbf041)

azat commented:
- `test_insert_quorum_with_keeper_loss_connection` is flaky - https://github.com/ClickHouse/ClickHouse/issues/67604, but [this time](https://s3.amazonaws.com/clickhouse-test-reports/71406/8b3ce129456a1f85839a48538780639e2e3c3020/integration_tests__asan__old_analyzer__[6_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel3_0.log) it failed with a different reason, which will be fixed here - https://github.com/ClickHouse/ClickHouse/pull/71418

azat commented:
>[Stateless tests (ubsan) [2/2] — Server died, fail: 17, passed: 388, skipped: 5](https://s3.amazonaws.com/clickhouse-test-reports/71406/2c52da92feb038da8ae0dc9e37cf4eb30f8b3128/stateless_tests__ubsan__[2_2].html)

```
/var/log/clickhouse-server/clickhouse-server.err.log:2024.11.04 09:31:07.829197 [ 1264 ] {} <Fatal> : Logical error: 'Part all_61_61_0 from table test_j2lmd2kq.alter_table1 (b42bc2ca-2251-436a-990e-3104aa0976cb) remains in ZooKeeper after DROP_RANGE all_0_61_999999999_999999999'.
```

- https://github.com/ClickHouse/ClickHouse/issues/56037

azat committed:
Automatic ORDER BY to disk (max_bytes_ratio_before_external_sort)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>Conflicting files
>src/Core/SettingsChangesHistory.cpp

Fixed and also added separate setting for `ORDER BY`.

azat commented on src/Interpreters/Aggregator.cpp:
```diff
@@ -1498,13 +1501,25 @@ bool Aggregator::executeOnBlock(Columns columns,
+        bool write_temporary_file = false;
+        if (params.max_bytes_before_external_group_by && current_memory_usage > static_cast<Int64>(params.max_bytes_before_external_group_by))
+            write_temporary_file = true;
+        if (params.max_bytes_ratio_before_external_group_by > 0. && current_memory_usage > memory_limit * params.max_bytes_ratio_before_external_group_by)
```
It does make sense, however I think that per-query ratio can also be useful, since this will allow to install it to i.e. `0.5` and this will allow to avoid `MEMORY_LIMIT_EXCEEDED` errors, while it does not make a lot of sense to set ratio for memory available in total to `0.5`

So I have two options then:
- add server setting, i.e. `max_bytes_ratio_before_external_group_by_for_server`/`max_bytes_ratio_before_external_sort_for_server`
- or make this limits to trigger on the total memory consumption as well, i.e. if it was set to `0.5` then it will start using disk once either query memory limits reached 50% or the server limit

Thoughts?

azat replied:
Added two new server settings to control this behavior server-wide.

Michicosun replied:
I was actually thinking about finding the difference between user memory tracker hard limit and user memory usage.

Seems like it will exactly memory_limit * params.max_bytes_ratio_before_external_group_by if there is no queries running right now and automatically cover case of multiple queries.

azat committed:
Implement max_bytes_ratio_before_external_group_by_for_server

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Implement max_bytes_ratio_before_external_sort_for_server

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Cover max_bytes_ratio_before_external_{order,group_by}_for_server

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: disable test_max_bytes_ratio_before_external_order_group_by_for_server under TSan

The problem is that TSan has some memory overhead which increases
process RSS to 10Gi in the middle of the query and it will fail.
Another option is to avoid syncing with RSS too frequently, but I doubt
that it is significant to run this test under TSan

v1: increase memory to 20Gi
    Fixes: https://s3.amazonaws.com/clickhouse-test-reports/71406/b33be86dee3c7616e9193c339837b9e250810557/integration_tests__tsan__[6_6].html
v2: instead just disable under TSan
    Fixes: https://s3.amazonaws.com/clickhouse-test-reports/71406/df68cf7362d825c3d83991fd74c391a49e73a8b1/integration_tests__tsan__[6_6].html

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
Wow, green CI, haven't seen this for awhile

Michicosun commented on src/Processors/Transforms/MergeSortingTransform.cpp:
```diff
@@ -166,7 +186,33 @@ void MergeSortingTransform::consume(Chunk chunk)
+    if (!external_sort && max_bytes_ratio_before_external_sort > 0.)
+    {
+        auto [current_memory_usage, memory_limit] = getCurrentQueryMemoryUsageAndLimit();
+        if (current_memory_usage > memory_limit * max_bytes_ratio_before_external_sort)
```
here probably is better to use `sum_bytes_in_blocks`

azat replied:
Hm, but it does not takes into account memory usage in other places, i.e. aggregation

Michicosun reviewed (changes requested):
I think we can merge per query setting(`max_bytes_ratio_before_external_group_by`), because it is a sugar extension for currently used `max_bytes_before_external_group_by`.

Or use the method suggested in the issue and compute the memory limit for the query once before it starts as `ratio_setting * remaining_memory_amount` and use it similarly to `max_bytes_before_external_group_by`. In this case, to compute `remaining_memory_amount`, we need to find the memory tracker in the hierarchy with the most strict hard limit and check its occupancy.

azat commented on docs/en/operations/server-configuration-parameters/settings.md:
```diff
@@ -491,6 +491,26 @@ Type: Double
 
 Default: 0.9
 
+## max_bytes_ratio_before_external_group_by_for_server
```
>This is a dangerous setting

Why do you find it dangerous? It can be enabled only by admins

> I'm not sure it's necessary and probably shouldn't be enabled by default
> It actually enables degradation of all requests on the server.

It is not enabled yet anyway, the main idea to enable it by default (later), is to avoid MEMORY_LIMIT_EXCEEDED errors, but I'm not insist here

azat commented:
> I think we can merge per query setting(max_bytes_ratio_before_external_group_by), because it is a sugar extension for currently used max_bytes_before_external_group_by.

How do you think `max_bytes_ratio_before_external_group_by` can be merged with `max_bytes_before_external_group_by`?

I agree that it is kind of configuration sugar, but not completely.

The difference is that this ratio is supported all the time, not only at the query start (like described in #69286, and that 10% of cases that is not covered by that proposal is that are covered by this patches), and I do find it useful, imagine a situation when the user spawned 100 queries at one point in time, that will eventually requires more RAM that the node has.

> Or use the method suggested in the issue and compute the memory limit for the query once before it starts as ratio_setting * remaining_memory_amount and use it similarly to max_bytes_before_external_group_by. In this case, to compute remaining_memory_amount, we need to find the memory tracker in the hierarchy with the most strict hard limit and check its occupancy.

Again, I'm not really like this idea, because it does allow to cover the case with query peaks.

And I actually looked through more advanced functionality that had been described in #41887, but I'm not sure that the complexity described there worth it, with taking into respect that this patches covers 99% cases.

_P.S. I also followed the same convention as some other settings, like:_
- _`remerge_sort_lowered_memory_bytes_ratio`_
- _`memory_overcommit_ratio_*`_
- _`max_server_memory_usage_to_ram_ratio`_

azat committed:
Properly handle unlimited memory limit for automatic GROUP/ORDER BY

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on src/Interpreters/Aggregator.cpp:
```diff
@@ -165,6 +166,14 @@ size_t getMinBytesForPrefetch()
+    /// Use query-level memory tracker
+    if (auto * memory_tracker_child = DB::CurrentThread::getMemoryTracker())
+        if (auto * memory_tracker = memory_tracker_child->getParent())
+            return std::make_pair(memory_tracker->get(), memory_tracker->getHardLimit());
```
Indeed, nice catch! Fixed.

azat commented on src/Processors/QueryPlan/SortingStep.cpp:
```diff
@@ -87,7 +96,7 @@ SortingStep::SortingStep(
     , sort_settings(settings_)
 {
-    if (sort_settings.max_bytes_before_external_sort && sort_settings.tmp_data == nullptr)
+    if ((sort_settings.max_bytes_before_external_sort || sort_settings.max_bytes_ratio_before_external_sort > 0.) && sort_settings.tmp_data == nullptr)
```
Fixed

azat committed:
Respect max_bytes_ratio_before_external_sort_for_server for ORDER BY storage check

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Respect max_bytes_ratio_before_external_sort_for_server in Aggregator::mergeOnBlock()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on src/Interpreters/Aggregator.cpp:
```diff
@@ -2911,13 +2936,25 @@ bool Aggregator::mergeOnBlock(Block block, AggregatedDataVariants & result, bool
+        bool write_temporary_file = false;
+        if (params.max_bytes_before_external_group_by && current_memory_usage > static_cast<Int64>(params.max_bytes_before_external_group_by))
+            write_temporary_file = true;
+        if (params.max_bytes_ratio_before_external_group_by > 0. && current_memory_usage > memory_limit * params.max_bytes_ratio_before_external_group_by)
```
Thanks, fixed.

azat committed:
Reduce copy-paste by introducing Aggregator::writeToTemporaryFileIfNeeded()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on src/Interpreters/Aggregator.cpp:
```diff
@@ -1498,13 +1501,38 @@ bool Aggregator::executeOnBlock(Columns columns,
     {
-        size_t size = current_memory_usage + params.min_free_disk_space;
-        writeToTemporaryFile(result, size);
+        bool write_temporary_file = false;
+        if (params.max_bytes_before_external_group_by && current_memory_usage > static_cast<Int64>(params.max_bytes_before_external_group_by))
+            write_temporary_file = true;
+        if (!write_temporary_file && params.max_bytes_ratio_before_external_group_by > 0. && current_memory_usage > memory_limit * params.max_bytes_ratio_before_external_group_by)
+        {
+            LOG_TEST(log, "Use external aggregation due to max_bytes_ratio_before_external_group_by reached. Current memory usage is {} (> {}*{})",
+                formatReadableSizeWithBinarySuffix(current_memory_usage),
+                formatReadableSizeWithBinarySuffix(memory_limit),
+                params.max_bytes_ratio_before_external_group_by);
+            write_temporary_file = true;
+        }
+        if (!write_temporary_file && params.max_bytes_ratio_before_external_group_by_for_server > 0.)
+        {
+            auto server_memory_usage = total_memory_tracker.get();
+            auto server_memory_limit = total_memory_tracker.getHardLimit();
+            if (server_memory_usage > server_memory_limit * params.max_bytes_ratio_before_external_group_by_for_server)
+            {
+                LOG_TEST(log, "Use external aggregation due to max_bytes_ratio_before_external_group_by_for_server reached. Current server memory usage is {} (> {}*{})",
+                    formatReadableSizeWithBinarySuffix(server_memory_usage),
+                    formatReadableSizeWithBinarySuffix(server_memory_limit),
+                    params.max_bytes_ratio_before_external_group_by_for_server);
+                write_temporary_file = true;
+            }
+        }
+
+        if (write_temporary_file)
+        {
+            size_t size = current_memory_usage + params.min_free_disk_space;
+            writeToTemporaryFile(result, size);
+        }
```
Done.

Michicosun commented:
> How do you think `max_bytes_ratio_before_external_group_by` can be merged with `max_bytes_before_external_group_by`?

I meant to merge the changes with this setting (`max_bytes_ratio_before_external_group_by`).

> I agree that it is kind of configuration sugar, but not completely.

If you set `max_bytes_before_external_group_by` = `max_bytes_ratio_before_external_group_by` * `memory_limit` then it will have the same logic you wrote for `max_bytes_ratio_before_external_group_by`.

But I agree that it is more convenient than the previous one.

> And I actually looked through more advanced functionality that had been described in #41887, but I'm not sure that the complexity described there worth it, with taking into respect that this patches covers 99% cases.

Right now you implemented 2 new types of settings:
1. `max_bytes_ratio_*` - this is by complexity and solving problems is the same as `max_bytes_before_*`
2. `*_for_server` - this adds new behaviour. If server using more memory than this setting forces to spill data to disk of every running query.

Obviously 1-st can't solve #69286 because it has old logic. 

Second on the other hand affects all queries at the same time - this is not per-query level setting. For example if limit is 0.7 and current memory consumption is 0.69 and you started new query it will be spilled to disk right after start, because total memory consumption will become > 0.7.

In proposed solution it is important to recalculate limit from free memory related to the current user. So why your solution handles 99% of cases?

azat commented:
Thanks, I guess now I understand your point and it does make sense, let's verify if I understood your correctly:

- leave only `max_bytes_ratio_before_external_group_by` (no server settings)
- use hard memory limit of the server for this ratio
- _use strictest hard memory limit for this ratio_

>I meant to merge the changes with this setting (max_bytes_ratio_before_external_group_by).

But it is still unclear what you are suggesting to merge

> In proposed solution it is important to recalculate limit from free memory related to the current user. So why your solution handles 99% of cases?

I haven't thought about this case, my point was that #41887 looks too complex to me (with an extra hook), and I think that it is better to try dumb implementation first (like this patch, but using per-user memory limits)

azat committed:
Move settings changes to 24.12

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Change logic of automatic ORDER/GROUP BY to be based on available memory

- remove max_bytes_ratio_before_external_{order,group_by}_for_server
- change the way max_bytes_ratio_before_external_{order,group_by} works

Note, that it is not enough to transform ratio to bytes in
executeQuery(), since in this case it will not work for merges and
internal queries, plus, you have to reset them for Distributed engine
and update it for Merge/View/...

This patch also introduce some helpers (see MemoryTrackerUtils) and
adjust Aggregator::Params constructor to accept Settings object instead
of tons of arguments.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

Michicosun commented on src/Interpreters/Aggregator.cpp:
```diff
@@ -175,6 +194,77 @@ Block Aggregator::getHeader(bool final) const
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Setting max_bytes_ratio_before_external_group_by should be >= 0 and < 1 ({})", ratio);
+
+        UInt64 available_system_memory = getMostStrictAvailableSystemMemory();
+        max_bytes_before_external_group_by = static_cast<size_t>(available_system_memory * ratio);
```
maybe it is better to not rely on uint64_t max here and if there is no limit on memory just log something and ignore this setting?

azat replied:
Make sense, thanks, fixed.

Michicosun commented on src/Interpreters/Aggregator.h:
```diff
@@ -96,103 +96,72 @@ class Aggregator final
+        /// Do we need to put into AggregatedDataVariants::without_key aggregates for keys that are not in max_rows_to_group_by.
+        const bool overflow_row = false;
+        const size_t max_rows_to_group_by = 0;
+        const OverflowMode group_by_overflow_mode = OverflowMode::BREAK;
```
It is better to use OverflowMode::Throw as default here, like was before.

Michicosun commented on src/Processors/QueryPlan/SortingStep.cpp:
```diff
@@ -54,6 +57,23 @@ SortingStep::Settings::Settings(const Context & context)
+        if (ratio < 0 || ratio >= 1.)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Setting max_bytes_ratio_before_external_sort should be >= 0 and < 1 ({})", ratio);
+
+        UInt64 available_system_memory = getMostStrictAvailableSystemMemory();
```
the same about limits

Michicosun reviewed (changes requested):
LGTM, just a few suggestions.

azat committed:
Fix copy-paste typo for max_bytes_ratio_before_external_group_by doc

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix default value for group_by_overflow_mode

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Ignore max_bytes_ratio_before_external_{order,group_by} if memory limit is not configured

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
I've rebased on top of upstream (to fix conflicts in settings changes), but the changes had been done on top for easier incremental review

Michicosun commented:
`test_storage_rabbitmq` https://github.com/ClickHouse/ClickHouse/issues/71049
